openpype/hosts/maya/plugins/create/create_multishot_layout.py

```diff
@@ -0,0 +1,206 @@
+        # go through the current folder path and add each part to the list,
+        # but mark the current folder.
+        for part_idx in range(len(current_path_parts)):
+            label = current_path_parts[part_idx]
+            if current_path_parts[part_idx] == current_folder["name"]:
+                label = f"{current_path_parts[part_idx]} (current)"
+            items_with_label.append(
+                dict(label=label,
+                     value="/".join(current_path_parts[:part_idx + 1]))
+            )
+        # add the project as the first item
+        items_with_label.insert(
+            0, dict(label=f"{self.project_name} "
+                          "(shots directly under the project)", value=""))
+
+        return [
+            EnumDef("shotParent",
+                    default=current_folder["name"],
+                    label="Shot Parent Folder",
+                    items=items_with_label,
+                    ),
+            BoolDef("groupLoadedAssets",
+                    label="Group Loaded Assets",
+                    tooltip="Enable this when you want to publish group of "
+                            "loaded asset",
+                    default=False),
+            TextDef("taskName",
+                    label="Associated Task Name",
+                    tooltip=("Task name to be associated "
+                             "with the created Layout"),
+                    default="layout"),
+        ]
+
+    def create(self, subset_name, instance_data, pre_create_data):
+        shots = self.get_related_shots(
+            folder_path=pre_create_data["shotParent"]
+        )
+        if not shots:
+            # There are no shot folders under the specified folder.
+            # We are raising an error here but in the future we might
+            # want to create a new shot folders by publishing the layouts
+            # and shot defined in the sequencer. Sort of editorial publish
+            # in side of Maya.
+            raise CreatorError("No shots found under the specified folder.")
+
+        # Get layout creator
+        layout_creator_id = "io.openpype.creators.maya.layout"
+        layout_creator: Creator = self.create_context.creators.get(
+            layout_creator_id)
+
+        # Get OpenPype style asset documents for the shots
+        op_asset_docs = get_assets(
+            self.project_name, [s["id"] for s in shots])
+        for shot in shots:
+            # we are setting shot name to be displayed in the sequencer to
+            # `shot name (shot label)` if the label is set, otherwise just
+            # `shot name`. So far, labels are used only when the name is set
+            # with characters that are not allowed in the shot name.
+            if not shot["active"]:
+                continue
+
+            # get task for shot
+            asset_doc = next(
+                asset_doc for asset_doc in op_asset_docs
+                if asset_doc["_id"] == shot["id"]
+
+            )
```

Comment:
It's much better to before the `for shot in shots:` loop to prepare the lookup to asset doc by id.

```python
asset_docs_by_id = {doc["_id"]: doc for doc in asset_docs}
```

Then here you would just do:
```python
asset_doc = asset_docs_by_id[shot["id"]
```

Reply:
true, did some tests, no discernible performance issue with few shots, but with thousands there might be noticable change (but I broke everything along the way)